src/derivatives/elementwise.jl

```diff
@@ -122,23 +123,54 @@ for (g!, g) in ((:map!, :map), (:broadcast!, :broadcast))
             return out
         end
     end
-    for A in ARRAY_TYPES, T in (:TrackedArray, :TrackedReal)
-        @eval function Base.$(g!)(f::ForwardOptimize{F}, out::TrackedArray{S}, x::$(T){X}, y::$A) where {F,S,X}
-            result = DiffResults.GradientResult(SVector(zero(S), zero(S)))
-            df = (vx, vy) -> ForwardDiff.gradient!(result, s -> f.f(s[1], s[2]), SVector(vx, vy))
+    for A in ARRAY_TYPES
+        @eval function Base.$(g!)(f::ForwardOptimize{F}, out::TrackedArray, x::TrackedReal{X,D}, y::$A) where {F,X,D}
+            result = DiffResults.DiffResult(zero(X), zero(D))
+            df = let result=result
+                (vx, vy) -> let vy=vy
+                    ForwardDiff.derivative!(result, s -> f.f(s, vy), vx)
+                end
+            end
             results = $(g)(df, value(x), value(y))
             map!(DiffResult.value, value(out), results)
             cache = (results, df, index_bound(x, out), index_bound(y, out))
-            record!(tape(x, y), SpecialInstruction, $(g), (x, y), out, cache)
+            record!(tape(x), SpecialInstruction, $(g), (x, y), out, cache)
             return out
         end
-        @eval function Base.$(g!)(f::ForwardOptimize{F}, out::TrackedArray, x::$A, y::$(T){Y}) where {F,Y}
-            result = DiffResults.GradientResult(SVector(zero(S), zero(S)))
-            df = (vx, vy) -> ForwardDiff.gradient!(result,  s -> f.f(s[1], s[2]), SVector(vx, vy))
+        @eval function Base.$(g!)(f::ForwardOptimize{F}, out::TrackedArray, x::$A, y::TrackedReal{Y,D}) where {F,Y,D}
+            result = DiffResults.DiffResult(zero(Y), zero(D))
+            df = let result=result
+                (vx, vy) -> let vx=vx
+                    ForwardDiff.derivative!(result, s -> f.f(vx, s), vy)
+                end
+            end
             results = $(g)(df, value(x), value(y))
             map!(DiffResult.value, value(out), results)
+            out = track(DiffResults.value.(results), D, tp)
```

Comment:
this function is inplace so the input `out` should be over-written?

Reply:
I believe the over-writing is happening in the previous line so this line seems unnecessary and potentially buggy

Reply:
Hmm good catch. I'm not sure, maybe this line ended up in the PR unintentionally. At least I don't remember why I added it. Let's see if anything breaks if I remove it.

Reply:
```suggestion
```